### ISSUE TITLE
CheckLoss.pm and CheckLatency.pm - incorrect evaluation of sample array (fix included)

### ISSUE THREAD

**Issue by rborgmaster** [oetiker/SmokePing#69]
Matchers CheckLoss and CheckLatency is evaluates the sample array using a foreach loop. This works OK under circumstances where the FPing sample array does not contain more values than the parameter x.

As stated in documentaiton, smokeping may pass more values than desried by the Length() method. If this happens, CheckLoss and CheckLatency will, instead of considering only the last x number of values, satisfy with any x number of values within the array that matches the l condition parameter.

Example:
The below loss pattern results in raised alert, while it really shouldn't have.
Pattern: CheckLoss(l=>25,x=>5)
Data: loss: 0%, 10%, 30%, 20%, 20%, 30%, 30%, 20%

**Comment by rborgmaster** [comment]
Reason for the misbehavior is that a foreach -loop is used to process all values of the array, instead of using a controlled for-loop that processes only the significant values.

**Comment by rborgmaster** [comment]
Patch for CheckLoss.pm

`48a49
> use List::Util qw(min max);
77c78,82
<     foreach $loss ( @{ $data->{loss} } ) {
---
>     my $x = min($self->{param}{x}, scalar @{ $data->{loss} });
> 
>     #Iterate thru last x number of samples, starting with the most recent
>     for (my $i=1;$i<=$x;$i++) {
>         $loss = $data->{loss}[$_-$i];
83c88
<             # Alert has already been raised.  Check to make sure ALL latencies in RTT are less than the target
---
>             # Alert has already been raised.  Evaluate and count consecutive loss values that are below threshold.
87c92
<             # Alert is not raised.  If all values are over the alert threshold OR unreachable, raise the alert
---
>             # Alert is not raised.  Evaluate and count consecutive loss values that are above threshold.`

**Comment by rborgmaster** [comment]
Patch for CheckLatency.pm
```
48a49
> use List::Util qw(min max);
78c79,83
<     foreach $rtt ( @{ $data->{rtt} } ) {
---
>     my $x = min($self->{param}{x}, scalar @{ $data->{rtt} });
> 
>     #Iterate thru last x number of samples, starting with the most recent
>     for (my $i=1;$i<=$x;$i++) {
>         $rtt = $data->{rtt}[$_-$i];
84c89
<             # Alert has already been raised.  Check to make sure ALL latencies in RTT are less than the target
---
>             # Alert has already been raised.  Evaluate and count consecutive rtt values that are below threshold.
88c93
<             # Alert is not raised.  If all values are over the alert threshold OR unreachable, raise the alert
---
>             # Alert is not raised.  Evaluate and count consecutive rtt values that are above threshold.
```

**Comment by rborgmaster** [comment]
I'm sorry but I can't. Partly because I don't know how but also because my employer have a very restrictive firewall which effectively blocks all sorts of file uploading. I hope my patch above will do and someone can merge it.

**Comment by rborgmaster** [comment]
I think I might actually have made it... but not sure :-) First time ever trying something with git and I'm not an experienced developer so bear with me. I'm not sure if I was supposed to change both files in one pull request but this is how it turned out when I edited the file via web browser.

**Comment by oetiker** [comment]
indeed! thanks

**Comment by Strykar** [comment]
@oetiker Can this issue be closed as the PR is merged?
